atum/src/main/scala/za/co/absa/atum/utils/controlmeasure/ControlMeasureBuilder.scala

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2018-2019 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.atum.utils.controlmeasure
+
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions.{abs, col, crc32, sum}
+import org.apache.spark.sql.types.{DecimalType, LongType, NumericType}
+import org.slf4j.LoggerFactory
+import za.co.absa.atum.core.ControlType
+import za.co.absa.atum.model.CheckpointImplicits.CheckpointExt
+import za.co.absa.atum.model.{Checkpoint, ControlMeasure, ControlMeasureMetadata, Measurement}
+import za.co.absa.atum.utils.controlmeasure.ControlMeasureUtils.getTimestampAsString
+
+import scala.util.Try
+
+trait ControlMeasureBuilder {
+  def withSourceApplication(sourceApplication: String): ControlMeasureBuilder
+  def withInputPath(inputPath: String): ControlMeasureBuilder
+  def withReportDate(reportDate: String): ControlMeasureBuilder
+  def withReportVersion(reportVersion: Int): ControlMeasureBuilder
+  def withCountry(country: String): ControlMeasureBuilder
+  def withHistoryType(historyType: String): ControlMeasureBuilder
+  def withSourceType(sourceType: String): ControlMeasureBuilder
+  def withInitialCheckpointName(initialCheckpointName: String): ControlMeasureBuilder
+  def withWorkflowName(workflowName: String): ControlMeasureBuilder
+
+  def build: ControlMeasure
+}
+
+
+object ControlMeasureBuilder {
+  /**
+   * Get builder instance
+   *
+   * @param df               dataframe to build ControlMeasure from
+   * @param aggregateColumns Columns to be aggregated, must be part of `ds`
+   * @return ControlMeasureBuilder to continue with
+   */
+  def forDF(df: DataFrame, aggregateColumns: Seq[String]): ControlMeasureBuilder =
```

Comment:
`aggregateColumns` always used to be optional. Thinking if it should be here or maybe another  `withAggregateColumns` or default it to empty

Reply:
Oh, it looked sort of as this is required as there was no default value in the original method signature (I would expect `aggregateColumns: Seq[String] = Seq.empty`).

But sure, if this is truly optional, I agree with your suggestion.

Reply:
Remember PWX to Cobrix. At most half of the datasets had aggregations.

Reply:
True, done. `aggregateColumns` is now an optional field settable by `withAggregateColumns` on the builder.